File: app/javascript/channels/room_channel.js
import consumer from "channels/consumer"

const appRoom = consumer.subscriptions.create("RoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    return alert(data['message']);
  },

  speak: function(message) {
    return this.perform('speak', {message: message});
  }
});

window.document.onkeydown = function(e) {
  if (e.key === 'Enter') {
    appRoom.speak(e.target.value);
    e.target.value = '';
    e.preventDefault();
  }
};

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  # メッセージを送信する側(Publisher)が、特定のメッセージを受信する側(Subscriber)
  # クライアントとは？
  #クライアントとは、IT分野におけるコンピュータ間のデータ通信の役割であり、データを提供される側の役割をするコンピュータあるいはソフトウェアです。

  # 購読とよばれる
  # 購読(サブスクライバされるとこのコードが呼び出される)
  # クライアントの接続が行われた際に実行されるメソッド
  # stream_fromメソッドはAction Cableが保持しているメソッドの1つで、データの送受信を行うことができます。
  # room_channel間（room_channel.rbとroom_channel.js）でデータを送受信することができます。
  def subscribed
    stream_from "room_channel"
  end

  # 購読してない
   # クライアントの接続が解除された際に実行されるメソッド
  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # room_channel.jsのreceivedメソッドにdata[‘message’]キーワード引数で送信しています。
  def speak(data)
    ActionCable.server.broadcast 'room_channel',{ message: data['message'] }
  end
end
